### PR TITLE
feat(transactions): allow future-dated transactions + recurring auto-post

### LINKED ISSUE #1080: Permit to create budget and trasaction for future
Sometimes we know a transaction will occur in 4 days, and we are not able to write it, since futur data are grayed out

Same for budget, having to wait the 1st is annoying, it's more logic to make your budget before the month start.

### PR DESCRIPTION
**Pull request by Khaostica**
## Summary
- Allow future-dated transactions everywhere they were previously gated by `max: Date.current` on the date input. Future entries appear in the activity feed with a Scheduled badge and don't affect today's balance until the date arrives.
- Add an `auto_post` flag on `RecurringTransaction` that, when enabled, triggers a daily cron job to materialize a real `Entry + Transaction` on `next_expected_date` and advance the recurring forward.
- Cap `Balance::ForwardCalculator` at `Date.current` so future entries don't silently push the displayed "current balance" into the future.

Closes #1080

### Phase 1 — Future-dated one-time transactions
- Dropped `max: Date.current` from the three transaction date inputs (`transactions/_form.html.erb`, `transactions/show.html.erb`, `transactions/bulk_updates/new.html.erb`). The `Entry` model itself already allowed future dates — the only block was the HTML constraint.
- `Balance::ForwardCalculator#calc_end_date` now caps at `Date.current`. Without this, a future-dated entry would extend the balance series past today and `account.balance` (which reads the latest `balances` row) would show the projected balance after the scheduled entry posts.
- Added a "Scheduled" badge in `transactions/_transaction.html.erb`, rendered next to the existing Pending / Duplicate / Split badges using a `calendar-clock` icon. New i18n keys at `transactions.transaction.scheduled` + `scheduled_tooltip`.

### Phase 2 — Recurring auto-post
- Migration adds `auto_post:boolean default false` to `recurring_transactions` plus a partial index on `(auto_post, next_expected_date)` filtered to `status = 'active' AND auto_post = true` so the daily job query stays fast as the table grows.
- `app/models/recurring_transaction/auto_poster.rb` — PORO under `app/models/` per Convention 2. Materializes a real `Entry` for one due, active, non-transfer recurring, then calls the existing `record_occurrence!` to advance `next_expected_date` forward. Idempotency: a second call in the same day sees the advanced date and returns `:skipped_not_due`. Returns a `Result` struct so the job can log skip reasons without raising.
- `app/jobs/post_due_recurring_transactions_job.rb` — iterates eligible rows, calls the AutoPoster, isolates per-row failures (one bad recurring doesn't break the batch; Sentry catches it).
- `config/schedule.yml` — new `post_due_recurring_transactions` entry, daily at 06:00 UTC, queue `scheduled`. Sits alongside the other `clean_*` / `refresh_*` cron jobs.
- `RecurringTransactionsController#toggle_auto_post` + route + UI toggle on the index row (success-colored `calendar-check` when on, neutral `calendar-clock` when off). Same shape as the existing `toggle_status` action.

### Why the budget half is already done
`Budget.latest_valid_budget_start_date` is `Date.current.beginning_of_month + 2.years` on `main`, so creating budgets for future months already works. The reporter filed this in October — that half got fixed silently. No budget changes in this PR.

## Out of scope

- **Recurring transfers.** The AutoPoster returns `:skipped_transfer` for any recurring with a `destination_account_id`. A transfer needs paired inflow/outflow entries + a `Transfer` row; that's its own coordination problem and not in sokie's V1 ask. Tests cover the skip path so the behavior is explicit. Worth a follow-up that calls `Transfer::Creator`.
- **Phase 3 / projected cashflow.** sokie noted this as "(Optionally)". The data primitives this PR ships (future entries + recurring next_expected_date) are exactly what a cashflow projection chart would consume. Deferred to a follow-up so this PR stays focused.
- **Catch-up posting.** If the worker is down for N days, the AutoPoster posts one entry on the missed date and advances forward by one occurrence. Posting all N missed occurrences would surprise users who'd reconcile manually in the meantime; the conservative single-post matches what most ledger apps do.

## Validation

This worktree is on Windows without a local Ruby toolchain, so `bin/rubocop`, `bundle exec erb_lint`, `bin/brakeman`, and `bin/rails test` could not run locally. CI will run the full matrix on the PR:

- `lint` — `bin/rubocop -f github`
- `lint_js` — `npm run lint` (no JS touched)
- `scan_ruby` — `bin/brakeman --no-pager`
- `test_unit` — `bin/rails test` (includes 17 new tests across the AutoPoster, the job, the controller toggle, the forward calculator cap, and the Entry future-date allowance)
- `test_system` — `DISABLE_PARALLELIZATION=true bin/rails test:system`
- `pipelock` — secret + agent-security diff scan

Manual checks in this worktree:
- Followed `CONTRIBUTING.md` (`Closes` keyword, fork-first, branch off `main`) and `.cursor/rules/project-conventions.mdc`. PORO under `app/models/recurring_transaction/` per Convention 2 (no `app/services/`). No new gem dependency. `Current.user` / `Current.family` everywhere. Semantic design tokens.
- Confirmed `Entry` model has no max-date validation — the HTML form cap was the only block.
- Confirmed the new partial-index syntax matches existing `idx_recurring_txns_*` indexes in `db/schema.rb`.
- Confirmed `entries.source` is a real column (schema line 585), so the AutoPoster's `source: "recurring_auto_post"` tag is valid for future querying / auditability.

## Notes
- No public API endpoints, rswag specs, or OpenAPI changes.
- English-only locale entries, mirroring how `ds.*` and `admin.invitations.*` are scoped here. Non-English locales fall back to English.
- Migration is non-destructive: adds a column with a default, no data backfill needed. Down-migration drops the column and the partial index.
- The PR includes a phasing comment on issue #1080 saying Phase 3 (projection) will follow as a separate PR — keeps sokie's expectations honest.

Refs: #1080

## Summary by CodeRabbit

* **New Features**
  * Per-recurring toggle to enable/disable auto-posting; daily scheduled job posts due auto-posted recurrings (transfer recurrings cannot be enabled)

* **UI**
  * Auto-post toggle in recurring list
  * New "scheduled (future-dated)" badge on transactions
  * Date fields now allow future dates

* **Localization**
  * New auto-post and scheduled transaction UI strings

* **Bug Fixes**
  * Forward balance calculation no longer projects past today

* **Tests / Chores**
  * Added tests, DB migration and schedule entry for auto-posting

[Review Change Stack](https://app.coderabbit.ai/change-stack/we-promise/sure/pull/1909?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by Khaostica** [comment]
Fixed in e48f95b9 — `toggle_auto_post` is now `post`-only:

```ruby
post :toggle_auto_post
```

The view's `link_to ... data: { turbo_method: :post }` already sends POST, so no UI breakage; Rails' default `protect_from_forgery` covers CSRF on the new endpoint.

Two notes on the scope of the fix:

1. The existing `match :toggle_status, via: [ :get, :post ]` in the same `member` block has the same shape. I left it alone because fixing it touches an unrelated controller action and pulls this PR off-topic. Worth a follow-up consistency PR — flagged in a code comment on the route.
2. On the secondary suggestion (replace `toggle_*` with explicit enable/disable endpoints): I deferred for the same scope reason — `:toggle_status` is the established pattern in this codebase and a consistency refactor wants its own PR. Happy to do that work as a follow-up if the maintainers want it.

File: app/models/recurring_transaction/auto_poster.rb [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  def call
+    return Result.new(status: :skipped_inactive, reason: "recurring transaction is not active") unless @recurring.active?
+    return Result.new(status: :skipped_not_due, reason: "next_expected_date is in the future") if @recurring.next_expected_date.future?
+    return Result.new(status: :skipped_no_account, reason: "recurring transaction has no source account") if @recurring.account_id.blank?
+    return Result.new(status: :skipped_transfer, reason: "recurring transfers are not auto-posted in V1") if @recurring.transfer?
+
+    entry = nil
+    RecurringTransaction.transaction do
+      entry = @recurring.account.entries.create!(
+        date: @recurring.next_expected_date,
```
[the remark this answers is not in the document]
Comment (Khaostica) [reply app/models/recurring_transaction/auto_poster.rb]: Fixed in 548a3011. Wrapped the critical section in `@recurring.with_lock` (`SELECT ... FOR UPDATE`) and re-check `active?` + `next_expected_date.future?` inside the lock. Pre-lock cheap-skip checks are kept so we don't grab a row lock just to exit on inactive / no-account / transfer rows.

Trace of the race that's now closed:

1. Worker A: pre-lock check passes (`active`, due today)
2. Worker B: pre-lock check also passes
3. Worker A: acquires `with_lock` (row-level)
4. Worker A: re-checks inside lock → still due → creates Entry → `record_occurrence!` advances `next_expected_date` → commits + releases lock
5. Worker B: blocked on `with_lock` until A releases
6. Worker B: re-loads via `lock!` → sees A's advanced `next_expected_date` → re-check returns `:skipped_not_due` → no Entry created

No deterministic concurrent-worker test added — Rails' `with_lock` is itself well-tested and stubbing it to simulate the race window would test implementation rather than behavior. The existing "second call in the same day skips" test exercises the same `next_expected_date`-advanced state path that the inside-lock re-check catches.

File: app/models/recurring_transaction/auto_poster.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        name: display_name,
+        amount: posting_amount,
+        currency: @recurring.currency,
+        entryable: Transaction.new(merchant_id: @recurring.merchant_id),
+        source: "recurring_auto_post"
+      )
+
+      entry.lock_saved_attributes!
+      entry.mark_user_modified!
```
[the remark this answers is not in the document]
Comment (Khaostica) [reply app/models/recurring_transaction/auto_poster.rb]: Fixed in 548a3011. Removed both `entry.lock_saved_attributes!` and `entry.mark_user_modified!` from the AutoPoster's post path.

You're right about the failure mode: `mark_user_modified!` sets `user_modified: true`, which makes `protected_from_sync?` true (per `Entry#protected_from_sync?` at line 300), which causes provider syncs to treat the row as protected — exactly the stale-data-next-to-provider-data scenario you described. I had cargo-culted the pattern from `TransactionsController#create`, but that flag is meant for *user-driven* form submissions, not system-generated rows.

The `source: "recurring_auto_post"` tag on the Entry is enough for any downstream code that needs to distinguish these from manual entries; no user-protection flag needed.

Added a regression test (`auto-posted entry is not marked user_modified`) asserting both `user_modified?` and `protected_from_sync?` stay false on the created Entry.

File: app/controllers/recurring_transactions_controller.rb [hunk 3]
```diff
@@ -61,6 +61,22 @@ def toggle_status
     end
   end
 
+  def toggle_auto_post
+    @recurring_transaction = Current.family.recurring_transactions.accessible_by(Current.user).find(params[:id])
+    @recurring_transaction.update!(auto_post: !@recurring_transaction.auto_post?)
+
+    message = @recurring_transaction.auto_post? ?
+      t("recurring_transactions.auto_post_enabled") :
+      t("recurring_transactions.auto_post_disabled")
```
[the remark this answers is not in the document]
Comment (Khaostica) [reply app/controllers/recurring_transactions_controller.rb]: Fixed in a78b18ec. Guarded the enable path on `toggle_auto_post`:

```ruby
if enabling && @recurring_transaction.transfer?
  flash[:alert] = t("recurring_transactions.auto_post_transfer_not_allowed")
  redirect_to recurring_transactions_path
  return
end
```

Disabling stays unguarded (covers any legacy rows where the flag could somehow be true), only enabling is blocked.

Added a controller test (`toggle_auto_post refuses to enable on a transfer recurring`) that:
1. Creates a transfer recurring (source + destination accounts)
2. Posts to the toggle endpoint
3. Asserts `flash[:alert]` matches the new translation and `auto_post` stays false

New i18n key: `recurring_transactions.auto_post_transfer_not_allowed`. The message explains *why* the toggle was refused and points at the deferral ("transfer auto-post support") rather than just rejecting silently.

File: app/models/recurring_transaction/auto_poster.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+      # Advance next_expected_date forward so the row is no longer "due"
+      # on subsequent job runs in the same day.
+      @recurring.record_occurrence!(@recurring.next_expected_date, posting_amount)
+    end
```
[the remark this answers is not in the document]
Comment (Khaostica) [reply app/models/recurring_transaction/auto_poster.rb:91]: Already addressed in 548a3011 — the review was generated against commit `616d30b430`, which predates the row-lock fix. Current `AutoPoster#call` runs the critical section under `@recurring.with_lock` and re-checks the due/active guards inside the lock:

```ruby
@recurring.with_lock do
  if !@recurring.active?
    skip_result = Result.new(status: :skipped_inactive, ...)
    next
  end
  if @recurring.next_expected_date.future?
    skip_result = Result.new(status: :skipped_not_due, ...)
    next
  end
  posted_entry = @recurring.account.entries.create!(...)
  @recurring.record_occurrence!(@recurring.next_expected_date, posting_amount)
end
```

`with_lock` is functionally identical to the suggested `transaction { @recurring.lock!; ... }` — both run `SELECT ... FOR UPDATE` followed by the block inside the same transaction. Race window between the cheap pre-lock check and the lock acquisition is closed by the in-lock re-check on `next_expected_date.future?`.

The pre-lock checks at lines 49–52 are kept deliberately for cheap-skip cases (inactive / no account / transfer) so we don't grab a row lock just to exit on those.

**Review by sure-design — COMMENTED** [review]
## DS Drift Patrol — 1 finding

**Rule 3 — Hand-rolled pill → `DS::Pill`**

`app/views/transactions/_transaction.html.erb`

The new "Scheduled" badge hand-rolls pill styles with `rounded-full px-1.5 py-0.5 border border-secondary text-secondary inline-flex items-center gap-1 text-xs font-medium`. Now that `DS::Pill` has landed, use the primitive instead:

```erb
<%= render DS::Pill.new(
  label: t("transactions.transaction.scheduled"),
  tone: :neutral,
  marker: false,
  icon: "calendar-clock"
) %>
```

This centralises pill styling in the design system so tone or sizing changes become a one-line edit across all callsites.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by jjmata** [comment]
## Review

This is a large, well-structured PR. The core model logic is solid. A few things worth checking before merge:

**`Balance::ForwardCalculator#calc_end_date`** — the `.min` cap is correct. Future-dated entries will appear in the activity feed with the "Scheduled" badge but won't push the balance series forward. The test that asserts no balance row has a future date is exactly the right contract to pin.

**`RecurringTransaction::AutoPoster`** — the double-check pattern inside `with_lock` is the right approach for preventing double-posts. One question: `entry.source = "recurring_auto_post"` — does the `entries` table already have a `source` column, or does this PR rely on another migration that isn't included? The schema diff doesn't show it being added, so presumably it already exists, but worth confirming the column is present before this lands.

**`toggle_auto_post` route** — using `post :toggle_auto_post` (POST-only) is the right call, and the comment explaining why `toggle_status` wasn't also fixed is appreciated. However, the view uses `data: { turbo_method: :post }` on a regular `link_to`. This works, but if JavaScript is disabled the link sends a GET, which will 404. This matches the existing `toggle_status` pattern so it's consistent, but it's worth noting.

**`toggle_auto_post` controller** — the `respond_to` block only handles `format.html`. If a Turbo Frame or Turbo Stream request ever hits this endpoint (e.g., the list is wrapped in a frame in the future), it will fall through to a 406. Not a current bug, but something to revisit when/if the recurring transactions list gets turbo-ified.

**Test concern — GET routing test:**
```ruby
test "toggle_auto_post does not respond to GET" do
  assert_raises(ActionController::RoutingError) do
    get toggle_auto_post_recurring_transaction_url(@recurring)
  end
end
```
In `ActionDispatch::IntegrationTest`, unroutable requests typically result in a `404` response rendered by the router middleware, *not* a raised exception — `ActionController::RoutingError` is raised internally and then caught by the middleware stack. This test may pass in development mode (where exceptions bubble) but silently pass for the wrong reason in test mode. Consider asserting `assert_response :not_found` instead, or use `assert_raises` only in unit tests that call the router directly.

**Cron timing:** The job runs at `6:00 AM UTC`. For users in UTC-6 or earlier, `Date.current` at 6 AM UTC is still the previous calendar day in their local timezone. Since `next_expected_date` is stored as a date without timezone, this should be fine — the job correctly asks "has this date arrived in UTC?" and the date field is timezone-agnostic.

**Missing: no test for `auto_post: true` on a `manual?` recurring with `expected_amount_avg: nil`** — what does `posting_amount` return? It falls through to `@recurring.amount`, which is correct, but an edge case worth a test case.

Overall the approach is well-reasoned and the test coverage is thorough. Addressing the GET test reliability issue would be the most important pre-merge fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Khaostica** [comment]
@jjmata I’ll address these concerns by 12:00 pm EST.

**Comment by Khaostica** [comment]
test(recurring): assert 404 on GET to toggle_auto_post + cover nil avg amount fallback (#1080)

Two follow-ups from the final maintainer review on #1909.

1. The GET-rejection test asserted `ActionController::RoutingError`, which is raised inside RouteSet and caught by the exceptions middleware in `ActionDispatch::IntegrationTest` — so the test would silently pass in test mode without actually exercising the 404 path. Switched to `assert_response :not_found` and added a read-back assertion that `auto_post` is unchanged.

2. Added a missing edge case: a manual recurring with `auto_post: true` flipped on before its first posted occurrence has  `expected_amount_avg = nil`. `posting_amount` must fall back to the template `amount`. Pins the fallback so a future tweak to the variance branch can't silently start nil-posting.

**Comment by sure-design** [comment]
## DS Drift Patrol — 1 finding

### R1 — DS Component Bypass
- `app/views/transactions/_transaction.html.erb` (~line 107): New "Scheduled" future-date indicator uses a raw `<span>` with pill-like styling instead of `DS::Pill`. Notably, this same PR correctly migrates all the other transaction badges (Pending, Review recommended, Possible duplicate, Split) to `DS::Pill` — the Scheduled badge should follow the same pattern.
  > `<span class="inline-flex items-center gap-1 text-xs font-medium rounded-full px-1.5 py-0.5 border border-secondary text-secondary" title="...">`

**Suggested fix** (matches the sibling badges in this same PR):
```erb
<%= render DS::Pill.new(
  label: t("transactions.transaction.scheduled"),
  tone: :neutral,
  marker: false,
  icon: "calendar-clock",
  title: t("transactions.transaction.scheduled_tooltip")
) %>
```

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Khaostica** [comment]
I’ll implement that fix by 3:00 pm EST

**Comment by Khaostica** [comment]
Migrated the Scheduled badge to `DS::Pill` in 3000c3a, which matches the sibling badges (Pending, Review recommended, Possible duplicate, Split) now that they've moved over on `main`.

**Comment by sure-design** [comment]
## DS Drift Patrol — 1 finding

`app/views/recurring_transactions/index.html.erb` · **Rule 1 — Hand-rolled tooltip via `title=` attribute**

```erb
link_to ..., title: auto_post_aria_label do
  icon(...)
end
```

Using a bare `title=` attribute for a tooltip bypasses `DS::Tooltip`. Browser-native tooltips have no consistent styling, are not keyboard-accessible, and do not appear on touch devices.

➜ Replace with:

```erb
render DS::Tooltip.new(text: auto_post_aria_label, as: :span) do
  icon(...)
end
```

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Khaostica** [comment]
I’ll implement this fix by 3:00 pm EST.

**Comment by Khaostica** [comment]
Replaced the bare title= with DS::Tooltip (as: :span) in 8cb47f7 — keyboard/touch-accessible, link's aria-label still provides the accessible name.

**Comment by Khaostica** [comment]
CI test_unit failures fixed in ccda494 — controller test now asserts the 404 (StoreLocation rescues RecordNotFound globally), and the AutoPoster/Job helpers use merchants(:one) so they don't collide with the netflix_subscription fixture under the new idx_recurring_txns_acct_merchant partial unique index.

**Comment by sure-design** [comment]
The PR description notes that `bin/rubocop`, `erb_lint`, `bin/brakeman`, and `bin/rails test` could not be run locally (Windows toolchain). Before this can move forward, please confirm all CI checks are green — particularly `test_unit` and `scan_ruby`. The 17 new tests are good, but we need a passing CI run before review.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by jjmata** [comment]
My biggest concern here is that future transactions cannot by definition have correct exchange rates.  How does the implementation address that @Khaostica or @dosu

**Comment by sure-design** [comment]
Three unresolved items from the automated review pass that need to be addressed before merge:

1. **DS drift** — The "Scheduled" badge in `app/views/transactions/_transaction.html.erb` uses a hand-rolled `<span>` with `rounded-full px-1.5 py-0.5 border border-secondary text-secondary`. All other status badges in this view use `DS::Pill`. Replace with `render DS::Pill.new(label: t("transactions.transaction.scheduled"), tone: :neutral, icon: "calendar-clock")`. (Flagged by sure-design bot.)

2. **Race condition in `AutoPoster#call`** — `PostDueRecurringTransactionsJob` runs on a schedule and could fire concurrently. Without a row-level lock, two workers can both post the same occurrence. Wrap the entry creation and `record_occurrence!` in `@recurring.with_lock { ... }`.

3. **Auto-post enabled for transfer recurrings** — `toggle_auto_post` unconditionally flips the flag, allowing it to be enabled on transfer-type recurrings where it's non-functional. Add a guard and return a clear flash alert if `@recurring_transaction.transfer?` and the new value would be `true`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Khaostica** [comment]
@jjmata dosu's breakdown is accurate. Since `ForwardCalculator` caps the balance series at `Date.current`, future entries never feed a displayed balance until its date arrives, and by then the daily rate import has the real rate (and auto-posted recurrings only materialize on the due date). The only gap was cosmetic: a foreign-currency future entry's converted display amount before its date. I've added an FX-specific tooltip on the Scheduled badge (`b5b885b`) flagging that the converted amount is provisional until the transaction date. No balance logic changed.

**Comment by Khaostica** [comment]
@jjmata this PR is ready for review.

**Comment by sure-design** [comment]
## DS Drift Patrol — update

Resolved 2 of 2 findings discovered across all runs on this PR.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sure-design** [comment]
This PR has resolved all prior DS-drift and maintainer review findings, but the branch currently shows `mergeable_state: dirty` (conflicts with main) and the real CI suite (lint/test_unit/test_system) has no results on the current head — only the security-scan bots have run. Could you rebase onto latest `main` and get a full CI run through (may need workflow approval as a fork PR) so this can get a final pass?

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sure-design** [comment]
## DS Drift Patrol — 1 finding

Note: this PR's two earlier DS Drift Patrol findings (hand-rolled "Scheduled" badge, hand-rolled tooltip on the recurring amount-range explainer) were resolved 2026-05-23/24 and confirmed all-clear 2026-07-23. This is a separate, previously-unflagged instance introduced in the same feature area.

### Rule 1 — Bypassing DS components (hand-rolled tooltip via `title=`)

**File:** `app/views/transactions/_transaction.html.erb:121`

```erb
<%= render DS::Pill.new(
  label: t("transactions.transaction.scheduled"),
  tone: :neutral,
  marker: false,
  icon: "calendar-clock",
  title: t(scheduled_tooltip_key)
) %>
```

`title:` renders straight into a native HTML `title=` attribute on `DS::Pill` (`app/components/DS/pill.html.erb:11`) rather than the design system's tooltip. Wrap this in `DS::Tooltip.new(text: t(scheduled_tooltip_key), as: :span)` around the pill instead — the same pattern already used correctly a few lines below for the auto-post toggle icon in this same PR.

---
*Scanned by DS Drift Patrol — diff-scope only (`+` lines); `app/components/DS/**`, generated assets, `test/`, `db/` excluded.*

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by jjmata** [comment]
There's work @winters27 is doing in this area, so copying him!

**Comment by winters27** [comment]
Context for the overlap: #3201 (merged) replaced the single next-date lens with materialized `recurring_occurrences` driven by `recurrence_rules` (weekly, nth-weekday, semimonthly, yearly, weekend adjustment, end modes), and #3202/#3203 build the surfaces on top. Two thoughts for when this revives:

- Phase 1 (future-dated one-time transactions and the ForwardCalculator cap) is independent of the recurring machinery and could land on its own.
- Phase 2's cron reads `calculate_next_expected_date`, which only understands day-of-month patterns and is now the legacy path. Materializing from `RecurringOccurrence#due_on` instead inherits every frequency plus weekend handling, and keeps one answer to what is due when. The schema also already carries an `autopay` flag on series, so auto-post could become that flag's behavior for manual accounts rather than a second toggle; the sync-and-nightly matching pass would then link the posted entry to its occurrence and close it as paid.

The features compose either way; whichever lands second rebases onto the other.

Happy to line this up on the bills side once #3202/#3203 land.

**Comment by jjmata** [comment]
> Happy to line this up on the bills side once #3202/#3203.

Let's see ... we need a good solution for exchange rates forecasting, how does recurring bills do that to estimate when a future bill is not in the family currency?

**Comment by winters27** [comment]
We do not guess future rates, and I do not think anyone should. A foreign bill keeps its own currency on its row, and the family-currency totals convert it at whatever the rate is right now. They re-convert every day, so by the time the bill is due the number has drifted to the real one, and the day it is paid it is the real one. No rate on file? The total says so instead of counting zero.

This PR does the same thing from its side: no rate is ever guessed or saved, and balances wait until the day arrives. So I do not think rate forecasting is a problem either feature has to solve. It only becomes one if we build long-range projection charts someday, and then it should be solved once in the rates layer so everything gets it.

They fit together fine. Bills already handles the future for anything that repeats: it knows what is coming and shows it everywhere, it just keeps it out of the ledger until the day it actually happens. What bills does not cover is a one-time future transaction, and that is exactly what the future-dating half here adds. Auto-post writes the actual transaction for manual accounts, which bills also leaves alone, and it would slot nicely on top of bills' scheduling when it revives. One catch worth flagging: on a bank-synced account an auto-posted transaction will meet its real imported copy later and double up, so the toggle probably wants to be for manual accounts.

What do you think?

**Comment by jjmata** [comment]
Yeah, and for future bills that makes sense (though it would be nice to know the impact of a EUR bill on my USD `Family` budget in case it breaks the bank!)

When people are entering future transactions in the hope of knowing their standing of their finances, it's quite deceiving if you don't at least use the last FX rate in the future conversion.  As things stand it probably defaults to a 1:1 FX rate, no?

**Comment by jjmata** [comment]
We can always re-open this if we like it, but I believe this was a Gittensor contributor that was not an active product user.  Please correct me if I am wrong @Khaostica. 🙏
